sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -222,24 +223,68 @@ object Hash {
     }
   }
 
-  // Transaction Builder class with builder methods to recursively encode nodes
-  private sealed class TransactionBuilderV1(
+  object TransactionMetadataBuilderV1 {
+    final case class Metadata(
+        actAs: ImmArray[Ref.Party],
+        commandId: Ref.CommandId,
+        submissionId: Ref.SubmissionId,
+        transactionUUID: UUID,
+        mediatorGroup: Int,
+        domainId: String,
+        ledgerEffectiveTime: Option[Time.Timestamp],
+        submissionTime: Time.Timestamp,
+        submissionSeed: Hash,
+        disclosedContracts: ImmArray[Node.Create],
```

Comment:
I'm a bit worried here that we're using the legacy hashing function for contract arguments here. IIRC we wanted to reversion input contracts according to the template they're being used with when we create the transaction submission. So the receiving nodes may not get exactly the same template argument as what we're signing here and then the validation fails later.

@remyhaemmerle-da What are your thoughts on this?

Reply:
Unfortunately the handling of upgradable disclosed contract is not completely fleshed out in 3.x
In the current implementation 
- The template ID of disclosed is the one of creation
- The protocol still requires the creation packages are loaded.
So I think it is fine for the first iteration of the hashing. 

I am afraid we will have to review quite a bit the hashing scheme when we cleanup upgrade in 3.x